content/blog/250826-redefining-cng-ccrp.md

```diff
@@ -0,0 +1,446 @@
+ranges above, we end up needing to retrieve 10 x 10 x 3652 = 365,200 chunks,
+each requiring an individual HTTP request. If the chunks were smaller to
+accommodate even more granular access, say 0.25° x 0.25° x 6 hours, then we end
+up having to make (4 x 4 x 4) x (10 x 10 x 3652) \= 23,372,800 requests!
+Assuming a rather optimistic time-to-first-byte of only 15 ms (running compute
+in region alongside the data, for example), we’d end up with almost 10 hours of
+time spent in the download process not transferring any bytes just because we
+have to make individual requests for each chunk.
+
+With CCRP, we can make one HTTP request to get all chunks, regardless of the
+chunk size:
+
+```plaintext
+GET https://ccrp.example.com/datasets/weather/data
+    ?time[gte]=315619200
+    &time[lt]=631152000
+    &lon[gte]=55
+    &lon[lt]=65
+    &lat[gte]=122
+    &lat[lt]=132
+    &variable=temperature
+```
+
+Admittedly, this request is for a whole lot of data, and it is unlikely a user
+would want to download all this with only one download stream. CCRP, as it is
+proposed, also has support for cases where users want or need more control over
+the download process, including progressive downloads to facilitate retries and
+parallel download streams for large amounts of data, as this case.
+
+### What could this look like with better tooling?
+
+With current tooling the original S3 request might look like this:
+
+```py
+import xarray as xr
+
+# Today - direct S3 access
+ds = xr.open_dataset(
+    's3://some_bucket/weather',
+    engine='zarr',
+)
+array = ds['temperature'].sel(
+    time=slice('2010-01-01', '2020-01-01'),
+    lat=slice(122, 132),
+    lon=slice(55, 65)
+).values  # to realize the values into an array, the
+          # tooling has to make a request for every chunk
+```
+
+Xarray and the lower-level tooling hides from the user that accessing the
+`values` attribute of the temperature array has to make a separate request for
+every chunk, as we’ve discussed.
+
+How could this same operation look using CCRP?
+
+```py
+# Tomorrow - same interface, just using CCRP
+ds = xr.open_dataset(
+    'ccrp://example.com/datasets/weather',
+    engine='zarr',
+)
+array = ds['temperature'].sel(
+    time=slice('2020-01-01', '2020-02-01'),
+    lat=slice(122, 132),
+    lon=slice(55, 65)
+).values  # the underlying tooling makes only one request
+          # to realize the array (or a few for parallelism)
+```
+
+This looks exactly the same, aside from the dataset URL. And that’s the point:
+CCRP is a low-level abstraction meant to make things easier for users, to plug
+the leaks in the current object model abstraction. It should be transparent to
+users.
+
+## What happens next
+
+I've drafted an [initial CCRP specification](https://ccrp.dev)
+that defines:
+
+* A RESTful API for listing, fetching metadata for, and querying datasets
+* A small but flexible query syntax for both array and tabular data
```

Comment:
So would CCRP be useful for parquet/iceberg?

Reply:
Yes, that's the idea. Raster and tabular data are remarkably similar at a conceptual level, they are just different ways of storing data in an n-dimensional space.